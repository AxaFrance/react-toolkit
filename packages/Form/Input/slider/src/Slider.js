import React from 'react';
import PropTypes from 'prop-types';
import Slider from 'rc-slider';
import 'rc-slider/assets/index.css';

import {
  InputConstants as Constants,
  Input,
  omit,
  withInput,
} from '@axa-fr/react-toolkit-form-core';

const omitProperties = omit([
  'classModifier',
  'className',
  'isVisible',
  'options',
]);

const mapOptions = (options) => options.map((o, index) => ({ index, ...o }));

const CustomSlider = (props) => {
  const {
    componentClassName,
    isVisible,
    onBlur,
    onFocus,
    disabled,
    readOnly,
    options,
    value,
    onChange,
    ...otherProps
  } = props;

  const newOptions = mapOptions(options);
  const indexes = newOptions.map((o) => o.index);

  const min = Math.min(...indexes);
  const max = Math.max(...indexes);

  const marks = {
    /* 0: 'VERY BAD',
      1: 'BAD',
      2: 'NICE',
      3: 'VERY NICE', */
  };
  newOptions.forEach((element) => {
    marks[element.index] = element.label;
  });

  let newValue = null;
  newOptions.forEach((element) => {
    if (element.value === value) {
      newValue = element.index;
    }
  });

  return (
    <Slider
      min={min}
      max={max}
      marks={marks}
      className={componentClassName}
      step={null}
      onChange={onChange}
      onAfterChange={onBlur}
      onBeforeChange={onFocus}
      value={newValue}
      disabled={disabled}
      readOnly={readOnly}
      {...omitProperties(otherProps)}
    />
  );
};

const handlers = {
  onChange: ({ name, id, onChange, options }) => (e) => {
    const newOptions = mapOptions(options);

    let option = null;
    newOptions.forEach((o) => {
      if (o.index === e) {
        option = o;
      }
    });

    onChange({
      value: option.value,
      name,
      id,
    });
  },
};

const propTypes = {
  ...Constants.propTypes,
  options: PropTypes.array.isRequired,
};

const defaultClassName = 'af-form__slider';
const defaultProps = {
  ...Constants.defaultProps,
};
const EnhancedComponent = withInput(
  defaultClassName,
  propTypes,
  defaultProps,
  handlers
)(CustomSlider);

EnhancedComponent.Clone = Input.Clone;
EnhancedComponent.displayName = Slider.name;

export default EnhancedComponent;
