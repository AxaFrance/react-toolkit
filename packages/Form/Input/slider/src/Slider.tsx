import React, { ComponentProps, ReactNode } from 'react';
import RcSlider from 'rc-slider';
import 'rc-slider/assets/index.css';

import {
  withComponentClassName,
  CustomFormEvent,
} from '@axa-fr/react-toolkit-form-core';

type Marks = Record<
  number,
  | React.ReactNode
  | {
      style?: React.CSSProperties;
      label?: string;
    }
>;

type Props = Omit<ComponentProps<typeof RcSlider>, 'marks' | 'onChange'> & {
  options: { value: number; label?: string | ReactNode }[];
  componentClassName?: string;
  onChange: (event: CustomFormEvent) => void;
  id: string;
  name: string;
};

const Slider = ({
  id,
  name,
  componentClassName,
  options,
  onChange,
  value,
  defaultValue,
  ...otherProps
}: Props) => {
  const optionSorted = [...options].sort(
    (optA, optB) => optA.value - optB.value
  );
  const marks = optionSorted.reduce(
    (prevMarks, { value: optValue, label: optLabel }, index) => {
      const newMarks = { ...prevMarks };
      newMarks[index] = optLabel || optValue;
      return newMarks;
    },
    {} as Marks
  );

  return (
    <RcSlider
      {...otherProps}
      min={0}
      max={optionSorted.length - 1}
      marks={marks}
      step={null}
      className={componentClassName}
      defaultValue={defaultValue || value}
      onChange={(e) => {
        onChange({
          value: optionSorted[e].value.toString(),
          name,
          id,
        });
      }}
    />
  );
};

export default withComponentClassName<Props>('af-form__slider')(Slider);
