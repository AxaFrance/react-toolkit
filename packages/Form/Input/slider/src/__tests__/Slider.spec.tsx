import React from 'react';
import { render } from '@testing-library/react';
import Slider from '../Slider';

describe('<Slider>', () => {
  it('renders Slider correctly', () => {
    const { asFragment } = render(
      <Slider id="id" name="file" options={[]} onChange={() => {}} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
