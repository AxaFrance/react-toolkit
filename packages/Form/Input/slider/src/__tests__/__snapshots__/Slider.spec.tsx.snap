// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Slider> renders Slider correctly 1`] = `
<DocumentFragment>
  <div
    class="rc-slider af-form__slider"
  >
    <div
      class="rc-slider-rail"
    />
    <div
      class="rc-slider-track"
      style="left: 0%; width: 0%;"
    />
    <div
      class="rc-slider-step"
    />
    <div
      aria-disabled="false"
      aria-valuemax="-1"
      aria-valuemin="0"
      aria-valuenow="0"
      class="rc-slider-handle"
      role="slider"
      style="left: 0%; transform: translateX(-50%);"
      tabindex="0"
    />
    <div
      class="rc-slider-mark"
    />
  </div>
</DocumentFragment>
`;
